--- isp-imx-4.2.2.22.0/imx/start_isp.sh.orig	2023-07-27 14:35:12.354472454 +0900
+++ isp-imx-4.2.2.22.0/imx/start_isp.sh	2023-07-27 15:29:32.889805778 +0900
@@ -10,6 +10,7 @@ RUNTIME_DIR="$( cd "$(dirname "$0")" >/d
 NR_DEVICE_TREE_BASLER=$(grep basler-camera-vvcam `find /sys/firmware/devicetree/base/soc@0/ -name compatible | grep i2c` -l | wc -l 2> /dev/null)
 NR_DEVICE_TREE_OV5640=$(grep ov5640 `find /sys/firmware/devicetree/base/soc@0/ -name compatible | grep i2c` -l | wc -l 2> /dev/null)
 NR_DEVICE_TREE_OS08A20=$(grep os08a20 `find /sys/firmware/devicetree/base/soc@0/ -name compatible | grep i2c` -l | wc -l 2> /dev/null)
+NR_DEVICE_TREE_OV5647=$(grep ov5647 `find /sys/firmware/devicetree/base/soc@0/ -name compatible | grep i2c` -l | wc -l 2> /dev/null)
 
 
 # check if the basler device has been enabled in the device tree
@@ -40,6 +41,17 @@ elif [ $NR_DEVICE_TREE_BASLER -eq 2 ]; t
 	# Default configuration for Basler daA3840-30mc: dual_basler_1080p60
 	# Available configurations: dual_basler_1080p60, dual_basler_1080p60hdr
 	exec ./run.sh -c dual_basler_1080p60 -lm
+
+# check if the ov5647 device has been enabled in the device tree
+elif [ $NR_DEVICE_TREE_OV5647 -eq 1 ]; then
+
+        echo "Starting isp_media_server for Single ov5647"
+
+        cd $RUNTIME_DIR
+        # Default configuration for OV5647: ov5647_1080p60
+        # Available configurations: ov5647_1080p60
+        exec ./run.sh -c ov5647_1080p60 -lm
+
 # check if the os08a20 device has been enabled in the device tree
 elif [ $NR_DEVICE_TREE_OS08A20 -eq 1 ]; then
 
--- isp-imx-4.2.2.22.0/imx/run.sh.orig	2023-07-27 14:35:12.358472468 +0900
+++ isp-imx-4.2.2.22.0/imx/run.sh	2023-07-27 15:24:22.524741637 +0900
@@ -34,6 +34,8 @@ USAGE+="\tos08a20_1080p30hdr      - sing
 USAGE+="\tdual_os08a20_1080p30hdr - dual os08a20 cameras on MIPI-CSI1/2, 1920x1080, 30 fps, HDR configuration\n"
 USAGE+="\tos08a20_4khdr           - single os08a20 camera on MIPI-CSI1, 3840x2160, 15 fps, HDR configuration\n"
 
+USAGE+="\tov5647_1080p60         - single ov5647 camera on MIPI-CSI1, 1920x1080, 60 fps\n"
+
 # parse command line arguments
 while [ "$1" != "" ]; do
 	case $1 in
@@ -87,6 +89,12 @@ write_default_mode_files () {
 	echo "[mode.3]" >> DAA3840_MODES.txt
 	echo "xml = \"DAA3840_30MC_1080P-hdr.xml\"" >> DAA3840_MODES.txt
 	echo "dwe = \"dewarp_config/daA3840_30mc_1080P.json\"" >> DAA3840_MODES.txt
+
+	# OV5647 modes file
+	echo -n "" > OV5647_MODES.txt
+        echo "[mode.0]" >> OV5647_MODES.txt
+        echo "xml = \"OV5647_8M_02_1080p_linear.xml\"" >> OV5647_MODES.txt
+        echo "dwe = \"dewarp_config/sensor_dwe_bypass_1080P_config.json\"" >> OV5647_MODES.txt
 }
 
 # write the sensonr config file
@@ -194,7 +202,7 @@ load_modules () {
 write_default_mode_files
 
 echo "Trying configuration \"$ISP_CONFIG\"..."
-MODULES_TO_REMOVE=("basler-camera-driver-vvcam" "os08a20" "ov2775" "${MODULES[@]}")
+MODULES_TO_REMOVE=("basler-camera-driver-vvcam" "os08a20" "ov2775" "ov5647" "${MODULES[@]}")
 case "$ISP_CONFIG" in
 		basler_4k )
 			MODULES=("basler-camera-driver-vvcam" "${MODULES[@]}")
@@ -308,6 +316,15 @@ case "$ISP_CONFIG" in
                          write_sensor_cfg_file "Sensor0_Entry.cfg" $CAM_NAME $DRV_FILE $MODE_FILE $MODE
                          write_sensor_cfg_file "Sensor1_Entry.cfg" $CAM_NAME $DRV_FILE $MODE_FILE $MODE
                          ;;
+		 ov5647_1080p60 )
+                         MODULES=("ov5647" "${MODULES[@]}")
+                         RUN_OPTION="CAMERA0"
+                         CAM_NAME="ov5647"
+                         DRV_FILE="ov5647.drv"
+                         MODE_FILE="OV5647_MODES.txt"
+                         MODE="0"
+                         write_sensor_cfg_file "Sensor0_Entry.cfg" $CAM_NAME $DRV_FILE $MODE_FILE $MODE
+                         ;;
 		 *)
 			echo "ISP configuration \"$ISP_CONFIG\" unsupported."
 			echo -e "$USAGE" >&2
